### PR TITLE
Refactor GS event notifications

### PR DESCRIPTION
**Pull request by kislyuk**
* Switch from a poor API (https://cloud.google.com/storage/docs/object-change-notification) to a better API (https://cloud.google.com/storage/docs/pubsub-notifications) for GS event notifications. Rip out the associated event forwarder in the `/internal/notify` API route. (The old API is impossible to automate from a clean slate account, and provides no way to list bucket notification subscriptions.)
* Add an event relay Google Cloud Function that forwards relevant GS events to SNS events.
* Add initial infrastructure for deploying Google Cloud Functions.
* Update deployment documentation.
* Add placeholder for indexer daemon GS event entry point.

### CODE REVIEW

File: scripts/deploy_gcf.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+try:
+    print(gcf_conn.api_request("POST", f"/{gcf_ns}", data=gcf_config))
+except google.cloud.exceptions.Conflict:
+    print(gcf_conn.api_request("PUT", f"/{gcf_ns}/{args.gcf_name}", data=gcf_config))
+
+sys.stderr.write("Waiting for deployment...")
+sys.stderr.flush()
+while gcf_conn.api_request("GET", f"/{gcf_ns}/{args.gcf_name}")["status"] == "DEPLOYING":
```
Comment (ttung) [inline scripts/deploy_gcf.py]: do we need a ceiling on how long this can wait?

File: scripts/deploy_gcf.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    sys.stderr.write(".")
+    sys.stderr.flush()
+    time.sleep(1)
+sys.stderr.write("done\n")
+
+res = gcf_conn.api_request("GET", f"/{gcf_ns}/{args.gcf_name}")
+print(res)
+assert gcf_conn.api_request("GET", f"/{gcf_ns}/{args.gcf_name}")["status"] == "READY"
+
+deploy_blob.delete()
```
Comment (ttung) [inline scripts/deploy_gcf.py]: this should be in a `finally` clause.

File: daemons/Makefile [hunk 3]
```diff
@@ -17,4 +17,7 @@ dss-chunked-task:
 	./build_deploy_config.sh $@ $(DSS_DEPLOYMENT_STAGE)
 	cd $@; domovoi deploy --stage $(DSS_DEPLOYMENT_STAGE)
 
-.PHONY: dss-chunked-task dss-sync dss-index
+dss-gs-event-relay:
+	$(DSS_HOME)/scripts/deploy_gcf.py $@ --entry-point "dss_gs_bucket_events_$(subst -,_,$(DSS_GS_BUCKET))"
```
Comment (ttung) [inline daemons/Makefile]: IIRC, `test_sync.py` uses the test bucket. this will deploy the script against the dev bucket.

Comment (kislyuk) [reply daemons/Makefile]: `test_sync.py` is a unit test - it doesn't call live lambdas, calling the daemon entry points directly in-process instead. So it shouldn't care about which bucket the deployed event bus is connected to.

File: scripts/deploy_gcf.py (hunk 1 above)
Comment (kislyuk) [reply scripts/deploy_gcf.py]: Will add.

File: scripts/deploy_gcf.py (hunk 2 above)
Comment (kislyuk) [reply scripts/deploy_gcf.py]: Will fix.

File: daemons/Makefile (hunk 3 above)
Comment (ttung) [reply daemons/Makefile]: Got it, thanks.

**Comment by kislyuk** [comment]
@ttung ready for another look.

**Comment by kislyuk** [comment]
Test strategy: I will write an integration test that gets called from a Travis cron job. This integration test will verify, in dev, sync and indexing of large bundles.

**Pull request merged**